### PR TITLE
refactor(loader): cache hash information

### PR DESCRIPTION
**Pull request by lewis6991**
Whenever we run fs_stat() on a path, save this information in the loader so it can be re-used.

- Loader.loadfile: Remove arguments `hash` as it is no longer needed.

- Loader.loader: Use _G.loadstring instead of Loader.load
  This allows plugins to wrap loadstring to inspection and profiling

### CODE REVIEW

**Comment by lewis6991** [comment]
@folke: will wait for your review on this.

The main goal here is for the loader to use `loadstring` so packer can profile plugins properly without via Lua builtin API's.

We are already maintaining a top-mod cache, so keeping a hot fs_stat cache seems like to good idea to me.

File: runtime/lua/vim/loader.lua [hunk 1]
```diff
@@ -41,6 +43,16 @@ local Loader = {
   },
 }
 
+--- @param path string
+--- @return CacheHash
+--- @private
+function Loader.get_hash(path)
+  if not Loader._hashes[path] then
+    Loader._hashes[path] = uv.fs_stat(path) --[[@as CacheHash]]
```
Comment (folke) [inline runtime/lua/vim/loader.lua:51]: Just a remark here, that we should never cache hashes for non-existing files. So the code as is is good, since it will be `nil` and getting the hash again will always redo an `fs_stat`. We need to keep doing it this way, or otherwise lazy will not be able to bootstrap properly in some cases.

TLDR: all good :)

File: runtime/lua/vim/loader.lua [hunk 2]
```diff
@@ -170,20 +192,6 @@ function Loader.loader_lib(modname)
...
 
---- `loadfile` using the cache
----@param filename? string
----@param mode? "b"|"t"|"bt"
----@param env? table
----@param hash? CacheHash
----@return function?, string?  error_message
----@private
--- luacheck: ignore 312
-function Loader.loadfile(filename, mode, env, hash)
```
Comment (folke) [inline runtime/lua/vim/loader.lua:181]: LuaJIT does support the mode and env parameters. See https://luajit.org/extensions.html

I'm actually using the `env` parameter in my personal configs to show virtual text where errors happened and with print statements when running a lua file.

The `env` parameter is useful when you want to sandbox or override the globals object when loading a lua module.

Removing this would break compatibility with LuaJIT, which is not something I think we want...

Comment (folke) [inline runtime/lua/vim/loader.lua:181]: Removing the entry point for the new global `loadfile` will also mess up the statistics.

Doing `require("whatever")` would now trigger stats for a regular load and additionally for loadfile, so not ideal.

File: runtime/lua/vim/loader.lua [hunk 3]
```diff
@@ -41,6 +43,16 @@ local Loader = {
   },
 }
 
+--- @param path string
+--- @return CacheHash
```
Comment (folke) [inline runtime/lua/vim/loader.lua:47]: nitpick, but you should probably return `uv_fs_t` here instead since that's what it is.
Might be useful later for other functionality.

The loader only needs the `CacheHash` which is a subset of `uv_fs_t`

File: runtime/lua/vim/loader.lua [hunk 4]
```diff
@@ -337,14 +340,17 @@ function M.find(modname, opts)
...
---- Resets the topmods cache for the path, or all the paths
+--- Resets the cache for the path, or all the paths
 --- if path is nil.
 ---@param path string? path to reset
 function M.reset(path)
   if path then
-    Loader._indexed[normalize(path)] = nil
+    path = normalize(path)
+    Loader._indexed[path] = nil
+    Loader._hashes[path] = nil
```
Comment (folke) [inline runtime/lua/vim/loader.lua:350]: This is not ideal, since `vim.loader.reset` expects a path on the rtp.

It's pretty much impossible to know all the lua mdules that would need to be cleared this way.

So either loop over all the hashes and clear those that start with the `path` parameter, or just always clear all hashes even when a `path` was given.

**Review by folke — CHANGES_REQUESTED** [review]
I'm fine with caching of the fs_stat calls, but removing compatibility with LuaJIT for loadfile is not ideal. Also see my comments on tracking stats for loadfile and load separately.

File: runtime/lua/vim/loader.lua (hunk 1 above)
Comment (lewis6991) [reply runtime/lua/vim/loader.lua:51]: I'll add a comment.

File: runtime/lua/vim/loader.lua (hunk 2 above)
Comment (lewis6991) [reply runtime/lua/vim/loader.lua:181]: Sounds good. I'll add it back and add a comment that luajit supports this and it is 5.1 compatible.

Comment (lewis6991) [reply runtime/lua/vim/loader.lua:181]: Ok, I've just reverted this.

File: runtime/lua/vim/loader.lua (hunk 4 above)
Comment (lewis6991) [reply runtime/lua/vim/loader.lua]: For now, I've just cleared all the hashes. Can potentially improve this in the future.

**Review by folke — APPROVED** [review]
Looks good to me! :)

File: runtime/lua/vim/loader.lua [hunk 5]
```diff
@@ -141,7 +165,7 @@ end
 function Loader.loader(modname)
   local ret = M.find(modname)[1]
   if ret then
-    local chunk, err = Loader.load(ret.modpath, { hash = ret.stat })
+    local chunk, err = loadfile(ret.modpath)
```
Comment (folke) [inline runtime/lua/vim/loader.lua:168]: Wait, we're still ging to track this twice. Should be Loader.load here?

**Review by folke — CHANGES_REQUESTED** [review]

File: runtime/lua/vim/loader.lua (hunk 4 above)
Comment (folke) [reply runtime/lua/vim/loader.lua]: I think that's good enough. Not worth improving this I think

File: runtime/lua/vim/loader.lua (hunk 5 above)
Comment (lewis6991) [reply runtime/lua/vim/loader.lua:168]: The goal of this PR is to use `_G.loadfile`, whatever that may be. This way user can wrap and profile `loadfile` themselves whether the cache is on/off.

Comment (folke) [reply runtime/lua/vim/loader.lua:168]: ok, no big deal.

**Review by folke — APPROVED** [review]
